src/page-storage/store-page.js

```diff
@@ -1,66 +1,97 @@
-    return {finalPage}
+    return finalPage
 }
 
-// Tries to reidentify the page in the tab, or creates a new page doc for it.
+// Tries to reidentify the page in tab, or creates a new page doc for it.
 // Either way, it returns the page doc. If a new page was created, it starts
-// page analysis and then deduplication, and also returns a promise
-// (finalPagePromise) of the page doc resulting from these steps.
-export async function reidentifyOrStorePage({tabId, url}) {
+// page analysis and then deduplication, and also returns an async function
+// (fetchFinalPage) returning the page doc resulting from these steps.
+export async function reidentifyOrStorePage({ tabId, url }) {
     // Find pages we know that had the same URL.
-    const samePageCandidates = (await findPagesByUrl({url})).rows.map(row => row.doc)
+    const samePageCandidates = (await findPagesByUrl({ url })).rows.map(row => row.doc)
 
     // Check if we can already tell in advance that this is the same page.
-    const reusablePage = await tryReidentifyPage({tabId, url, samePageCandidates})
+    const reusablePage = await tryReidentifyPage({ tabId, url, samePageCandidates })
 
     if (reusablePage) {
         // Equality is known (or assumed) in advance. Reuse the old page as is.
-        return {page: reusablePage}
+        return { page: reusablePage }
     } else {
-        // Create a new page doc in the database.
-        const page = await createPageStub({url})
-
-        // Start analysis and (possibly) deduplication, but do not wait for it.
-        const finalPagePromise = analyseAndTryDedupePage({
-            tabId,
-            page,
-            samePageCandidates,
-        })
-
-        // Return the page stub, and a promise of the analysed & deduped page.
-        return {page, finalPagePromise}
+        // Create a new page doc stub in the database.
+        const page = await storePage(createPage({ url }))
+
+        // Set up fetching function to run anaylsis + fetching logic on-demand
+        const fetchFinalPage = () => analyseAndTryDedupePage({ tabId, page, samePageCandidates })
+
+        // Return the page stub, and a function to get the analysed & deduped page.
+        return { page, fetchFinalPage }
+    }
+}
+
+/**
+ * Given a URL, attempts to perform all logic needed to store a full page document.
+ * @param {string} url The URL to attempt to store a page document for.
+ * @returns {any} Page document that was created (or found, if reusable/deduped).
+ */
+export async function storePageFromUrl(url) {
+    // Find pages we know that had the same URL.
+    const samePageCandidates = (await findPagesByUrl({ url })).rows.map(row => row.doc)
+
+    // Check if we can already tell in advance that this is the same page.
+    const reusablePage = await tryReidentifyPage({ url, samePageCandidates })
+
+    // Equality is known (or assumed) in advance. Reuse the old page as is.
+    if (reusablePage) {
+        return reusablePage
     }
```

Comment:
The thing I don't get is you have to update the references between visit docs (changing the `referringVisitId` gotten from browser model to an `_id` of another visit doc in our model). This needs all the visit docs available to do so, hence needs to happen right at the end before reindexing. Could still store visit docs directly at each step, but then the reference updating still needs to happen on completion.

>  I would probably still opt for importing and creating (nearly empty) page docs first

We could stick with this way, and batch them in another batching instance, if needed, but I just don't understand the purpose behind the page stubs. In page visit scenario, for a user it means the pages will show up in overview before all the data fetch/analyse stuff is done after the user visits the page (is that correct?), but in the importing history/bookmarks scenarios, what is the purpose?